core/src/main/java/org/infinispan/stream/impl/AbstractCacheStream.java

```diff
@@ -0,0 +1,734 @@
+   }
+
+   protected abstract S unwrap();
+
+   @Override
+   public boolean isParallel() {
+      return parallel;
+   }
+
+   boolean getParallelDistribution() {
+      return parallelDistribution == null ? true : parallelDistribution;
+   }
+
+   @Override
+   public S sequential() {
+      parallel = false;
+      return unwrap();
+   }
+
+   @Override
+   public S parallel() {
+      parallel = true;
+      return unwrap();
+   }
+
+   @Override
+   public S unordered() {
+      sorted = false;
+      return unwrap();
+   }
+
+   @Override
+   public S onClose(Runnable closeHandler) {
+      if (this.closeRunnable == null) {
+         this.closeRunnable = closeHandler;
+      } else {
+         this.closeRunnable = composeWithExceptions(this.closeRunnable, closeHandler);
+      }
+      return unwrap();
+   }
+
+   @Override
+   public void close() {
+      if (closeRunnable != null) {
+         closeRunnable.run();
+      }
+   }
+
+   <R> R performOperation(Function<S, ? extends R> function, boolean retryOnRehash, BinaryOperator<R> accumulator,
+                          Predicate<? super R> earlyTerminatePredicate) {
+      return performOperation(function, retryOnRehash, accumulator, earlyTerminatePredicate, true);
+   }
+
+   <R> R performOperation(Function<S, ? extends R> function, boolean retryOnRehash, BinaryOperator<R> accumulator,
+           Predicate<? super R> earlyTerminatePredicate, boolean ignoreSorting) {
+      // These operations are not affected by sorting, only by distinct
+      if (intermediateType.shouldUseIntermediate(ignoreSorting ? false : sorted, distinct)) {
+         return performIntermediateRemoteOperation(function);
+      } else {
+         ResultsAccumulator<R> remoteResults = new ResultsAccumulator<>(accumulator);
+         if (rehashAware) {
+            return performOperationRehashAware(function, retryOnRehash, remoteResults, earlyTerminatePredicate);
+         } else {
+            return performOperation(function, remoteResults, earlyTerminatePredicate);
+         }
+      }
+   }
+
+   <R> R performOperation(Function<S, ? extends R> function, ResultsAccumulator<R> remoteResults,
+                          Predicate<? super R> earlyTerminatePredicate) {
+      ConsistentHash ch = dm.getConsistentHash();
+      TerminalOperation<R> op = new SingleRunOperation<>(intermediateOperations,
+              supplierForSegments(ch, segmentsToFilter, null), function);
+      UUID id = csm.remoteStreamOperation(getParallelDistribution(), parallel, ch, segmentsToFilter, keysToFilter,
+              Collections.emptyMap(), includeLoader, op, remoteResults, earlyTerminatePredicate);
+      try {
+         R localValue = op.performOperation();
+         remoteResults.onCompletion(null, Collections.emptySet(), localValue);
+         try {
+            if ((earlyTerminatePredicate == null || !earlyTerminatePredicate.test(localValue)) &&
+                    !csm.awaitCompletion(id, 30, TimeUnit.SECONDS)) {
```

Comment:
Yes that is a TODO I had in the original list.  It is on ISPN-5626, I had it worded poorly.